src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.InFlightSolution.cs

```diff
@@ -100,29 +102,43 @@ public void IncrementInFlightCount_NoLock()
                 InFlightCount++;
             }
 
-            public void DecrementInFlightCount_NoLock()
+            /// <summary>
+            /// Returns the in-flight solution computations <em>when</em> the in-flight-count is decremented to 0. This
+            /// allows the caller to wait for those computations to complete (which will hopefully be quickly as they
+            /// will have just been canceled).  This ensures the caller doesn't return back to the host (potentially
+            /// unpinning the solution on the host) while the solution-computation tasks are still running and may still
+            /// attempt to call into the host.
+            /// </summary>
+            public ImmutableArray<Task<Solution>> DecrementInFlightCount_NoLock()
```

Comment:
i was considering having this return a Task that represented what the caller should await on.  however, we don't want this call awaited while in the state-lock, so that could be very confusing.  instead, we want this to run synchronously, clean up everything properly, then return the Task/Tasks the caller should await outsdie of hte lock.  So this is a bit clearer in that regard (imo).